conf_file.cpp

```diff
@@ -16,268 +16,416 @@
+    if (addr == MAP_FAILED)
+        return -errno;
 
-    file = fopen(_filename, "re");
-    if (!file) {
-        log_error_errno(errno, "Could not open conf file '%s' (%m)", _filename);
-        return -EIO;
+    _files = new conffile{fd, addr, (size_t)fstat.st_size, _files};
+
+    ret = _parse_file((char *)addr, (size_t)fstat.st_size, filename);
+    if (ret < 0) {
+        log_error("Conf file parcially parsed. Configuration keys defined before the error were "
+                  "included");
+        return ret;
     }
 
-    while ((read = getline(&entry, &n, file)) >= 0) {
+    return 0;
+}
+
+int ConfFile::_parse_file(const char *addr, size_t len, const char *filename)
+{
+    int line = 0, ret = 0;
+    char *end;
+    size_t line_len;
+
+    struct section *s = nullptr;
+    while ((end = (char *)memchr((void *)addr, '\n', len))) {
         line++;
+        line_len = end - addr;
 
-        _trim(entry);
+        // Make sure all casts to int are safe. No need to support options or sections
+        // larger than INT_MAX
+        if (line_len > INT_MAX)
+            return -EINVAL;
+        _trim(&addr, &line_len);
+        if (!line_len)
+            goto next;
 
-        switch (entry[0]) {
+        switch (addr[0]) {
         case ';':
         case '#':
-        case '\0':
-            // Discards comment or blank line
-            free(entry);
-            break;
+            goto next;
         case '[':
-            ret = _add_section(entry, line);
-            if (ret < 0) {
-                goto end;
-            }
+            s = _add_section(addr, line_len, line, filename);
+            if (!s)
+                return -EINVAL;
             break;
         default:
-            ret = _add_config(entry, line);
-            if (ret < 0) {
-                goto end;
+            if (!s) {
+                log_error(
+                    "On file %s: Line %d: Expected section before the definition of first option.",
+                    filename, line);
             }
-        }
 
-        entry = NULL;
-    }
-
-    _current_section = nullptr;
+            ret = _add_config(s, addr, line_len, filename, line);
+            if (ret < 0)
+                return ret;
+        }
 
-    if (!_sections || !_sections->configs) {
-        free(entry);
-        log_error("Invalid conf file %s", _filename);
-        return -EINVAL;
+    next:
+        len = len - (end - addr) - 1;
+        addr = end + 1;
     }
 
-end:
-    free(entry);
-    fclose(file);
     return ret;
 }
 
-const char *ConfFile::first_section()
+struct section *ConfFile::_add_section(const char *addr, size_t len, int line, const char *filename)
 {
-    _current_section = _sections;
-    return _current_section->name;
-}
+    const char *end;
+    struct section *s;
 
-const char *ConfFile::next_section()
-{
-    if (!_current_section->next)
-        return NULL;
+    end = (char *)memrchr(addr, ']', len);
```

Comment:
Hmmm... `memrchr` allows things like `[really ] weird] option]`. Is it desirable?